src/gz/bufread.rs

```diff
@@ -280,108 +298,213 @@ impl<R: BufRead + Write> Write for GzEncoder<R> {
+                => GzState::Header(buf),
+            Err(err) => GzState::Err(err)
+        };
 
-        let flate = deflate::bufread::DeflateDecoder::new(r);
         GzDecoder {
-            inner: CrcReader::new(flate),
-            header: Some(header),
-            finished: false,
+            inner: state,
+            reader: CrcReader::new(deflate::bufread::DeflateDecoder::new(r)),
+            multi: false,
+            header
         }
     }
 
-    fn finish(&mut self) -> io::Result<()> {
-        if self.finished {
-            return Ok(());
-        }
-        let ref mut buf = [0u8; 8];
-        {
-            let mut len = 0;
-
-            while len < buf.len() {
-                match self.inner.get_mut().get_mut().read(&mut buf[len..])? {
-                    0 => return Err(corrupt()),
-                    n => len += n,
-                }
-            }
-        }
-
-        let crc = ((buf[0] as u32) << 0)
-            | ((buf[1] as u32) << 8)
-            | ((buf[2] as u32) << 16)
-            | ((buf[3] as u32) << 24);
-        let amt = ((buf[4] as u32) << 0)
-            | ((buf[5] as u32) << 8)
-            | ((buf[6] as u32) << 16)
-            | ((buf[7] as u32) << 24);
-        if crc != self.inner.crc().sum() {
-            return Err(corrupt());
-        }
-        if amt != self.inner.crc().amount() {
-            return Err(corrupt());
-        }
-        self.finished = true;
-        Ok(())
+    fn multi(mut self, flag: bool) -> GzDecoder<R> {
+        self.multi = flag;
+        self
     }
 }
 
 impl<R> GzDecoder<R> {
     /// Returns the header associated with this stream, if it was valid
     pub fn header(&self) -> Option<&GzHeader> {
-        self.header.as_ref().and_then(|h| h.as_ref().ok())
+        self.header.as_ref()
     }
 
     /// Acquires a reference to the underlying reader.
     pub fn get_ref(&self) -> &R {
-        self.inner.get_ref().get_ref()
+        self.reader.get_ref().get_ref()
     }
 
     /// Acquires a mutable reference to the underlying stream.
     ///
     /// Note that mutation of the stream may result in surprising results if
     /// this encoder is continued to be used.
     pub fn get_mut(&mut self) -> &mut R {
-        self.inner.get_mut().get_mut()
+        self.reader.get_mut().get_mut()
     }
 
     /// Consumes this decoder, returning the underlying reader.
     pub fn into_inner(self) -> R {
-        self.inner.into_inner().into_inner()
+        self.reader.into_inner().into_inner()
     }
 }
 
 impl<R: BufRead> Read for GzDecoder<R> {
     fn read(&mut self, into: &mut [u8]) -> io::Result<usize> {
-        match self.header {
-            None => return Ok(0), // error already returned,
-            Some(Ok(_)) => {}
-            Some(Err(_)) => match self.header.take().unwrap() {
-                Ok(_) => panic!(),
-                Err(e) => return Err(e),
-            },
+        let GzDecoder { inner, header, reader, multi } = self;
+
+        enum Next {
```

Comment:
Having `enum Next` here makes it a bit hard to follow the control flow, especially `return Err` for wouldblock errors and a fall-through for other errors. Would it be possible to replace the current state with `None` on each iteration? That'd ideally remove the need for this `Next` and would also enable `?` to be used more uniformly in a few places.

Reply:
This is not as good as it is supposed to be, because we still need to set inner to correct state before throwing `WouldBlock`.